barretenberg/cpp/src/barretenberg/eccvm/transcript_builder.hpp

```diff
@@ -104,225 +130,397 @@ class ECCVMTranscriptBuilder {
+        const FF vm_x = vm_infinity ? 0 : vm_point.x;
+        const FF vm_y = vm_infinity ? 0 : vm_point.y;
+
+        // extract coordinates of the current accumulator
+        const FF accumulator_x = accumulator_infinity ? 0 : accumulator.x;
+        const FF accumulator_y = accumulator_infinity ? 0 : accumulator.y;
+
+        row.transcript_add_x_equal = (vm_x == accumulator_x) || (vm_infinity && accumulator_infinity);
+        row.transcript_add_y_equal = (vm_y == accumulator_y) || (vm_infinity && accumulator_infinity);
+
+        // compute the numerator and denominator of the slope
+        if ((accumulator_x == vm_x) && (accumulator_y == vm_y) && !vm_infinity && !accumulator_infinity) {
+            // Double case (x1 == x2, y1 == y2)
+            add_lambda_denominator = vm_y + vm_y;
+            add_lambda_numerator = vm_x * vm_x * 3;
+        } else if ((accumulator_x != vm_x) && !vm_infinity && !accumulator_infinity) {
+            // General case (x1 != x2)
+            add_lambda_denominator = accumulator_x - vm_x;
+            add_lambda_numerator = accumulator_y - vm_y;
         }
+    }
+    static void batch_invert(std::vector<FF>& inverse_trace_x,
+                             std::vector<FF>& inverse_trace_y,
+                             std::vector<FF>& transcript_msm_x_inverse_trace,
+                             std::vector<FF>& add_lambda_denominator,
+                             std::vector<FF>& msm_count_at_transition_inverse_trace,
+                             const size_t num_vm_entries)
+    {
         FF::batch_invert(&inverse_trace_x[0], num_vm_entries);
         FF::batch_invert(&inverse_trace_y[0], num_vm_entries);
         FF::batch_invert(&transcript_msm_x_inverse_trace[0], num_vm_entries);
         FF::batch_invert(&add_lambda_denominator[0], num_vm_entries);
         FF::batch_invert(&msm_count_at_transition_inverse_trace[0], num_vm_entries);
-        for (size_t i = 0; i < num_vm_entries; ++i) {
-            transcript_state[i + 1].base_x_inverse = inverse_trace_x[i];
-            transcript_state[i + 1].base_y_inverse = inverse_trace_y[i];
-            transcript_state[i + 1].transcript_msm_x_inverse = transcript_msm_x_inverse_trace[i];
-            transcript_state[i + 1].transcript_add_lambda = add_lambda_numerator[i] * add_lambda_denominator[i];
-            transcript_state[i + 1].msm_count_at_transition_inverse = msm_count_at_transition_inverse_trace[i];
-        }
+    }
+
+    static void finalize_transcript(std::vector<TranscriptRow>& transcript_state, const VMState& updated_state)
+    {
         TranscriptRow& final_row = transcript_state.back();
         final_row.pc = updated_state.pc;
         final_row.accumulator_x =
-            (updated_state.accumulator.is_point_at_infinity()) ? 0 : AffineElement(updated_state.accumulator).x;
+            updated_state.accumulator.is_point_at_infinity() ? 0 : AffineElement(updated_state.accumulator).x;
```

Comment:
Should probably add an explanation for (0,0)

Reply:
Currently, I don't have a good explanation, because that's exactly where the point at infinity breaks Goblin